diff --git a/src/lean.ml b/src/lean.ml
index ee9667f..9282aea 100644
--- a/src/lean.ml
+++ b/src/lean.ml
@@ -18,7 +18,9 @@ let __ () = assert false
 let invalid = Constr.(mkApp (mkSet, [| mkSet |]))
 
 let add_universe l ~lbound g =
-  let g = UGraph.add_universe l ~strict:false g in
+  let g =
+    UGraph.add_universe l ~lbound:UGraph.Bound.Set ~strict:false g
+  in
   UGraph.enforce_constraint (lbound, Le, l) g
 
 let quickdef ?(opaque = false) ~name ~types ~univs body =
@@ -134,7 +136,7 @@ let lean_scheme env ~dep mind u s =
     | Level s ->
       let (_, us), cst = uctx in
       assert (
-        Level.Set.cardinal us = 1 && PConstraints.is_empty cst);
+        Level.Set.cardinal us = 1 && Constraints.is_empty cst);
       let v = Level.Set.choose us in
       Vars.subst_univs_level_constr
         (Sorts.QVar.Map.empty, Level.Map.singleton v s)
@@ -217,17 +219,7 @@ let lean_scheme env ~dep mind u s =
     in
     Term.it_mkLambda_or_LetIn (Term.it_mkLambda_or_LetIn body fcs) paramsP
 
-let with_unsafe_univs f () =
-  let flags = Global.typing_flags () in
-  Global.set_typing_flags { flags with check_eliminations = false };
-  try
-    let v = f () in
-    Global.set_typing_flags flags;
-    v
-  with e ->
-    let e = Exninfo.capture e in
-    Global.set_typing_flags flags;
-    Exninfo.iraise e
+let with_unsafe_univs f () = f ()
 
 module N = LeanName
 
@@ -239,7 +231,7 @@ let univ_of_sort = function
   | Sorts.SProp -> None
   | Prop -> assert false
   | Set -> Some Universe.type0
-  | Type u | GSort (_, u) | VSort (_, u) -> Some u
+  | Type u | QSort (_, u) -> Some u
 
 let sort_max (s1 : Sorts.t) (s2 : Sorts.t) =
   match (s1, s2) with
@@ -250,8 +242,7 @@ let sort_max (s1 : Sorts.t) (s2 : Sorts.t) =
   | Set, Type u | Type u, Set ->
     Sorts.sort_of_univ (Univ.Universe.sup Univ.Universe.type0 u)
   | Type u, Type v -> Sorts.sort_of_univ (Univ.Universe.sup u v)
-  | GSort _, _ | _, GSort _ -> assert false
-  | VSort _, _ | _, VSort _ -> assert false
+  | QSort _, _ | _, QSort _ -> assert false
 
 (** [map] goes from lean names to universes (in practice either SProp or a named
     level) *)
@@ -377,7 +368,7 @@ let rec level_of_sets uconv n =
         else UnivGen.fresh_level ()
       in
       Global.push_context_set
-        (Level.Set.singleton l, UnivConstraints.singleton (p, Lt, l));
+        (Level.Set.singleton l, Constraints.singleton (p, Lt, l));
       sets := Int.Map.add n l !sets;
       let graph = add_universe l ~lbound:p uconv.graph in
       ({ uconv with graph }, l)
@@ -639,13 +630,13 @@ let univ_entry_gen { map; levels; graph } ounivs =
   let kept = Int.Map.fold (fun _ -> Level.Set.add) !sets kept in
   let csts = UGraph.constraints_for ~kept graph in
   let csts =
-    UnivConstraints.filter
+    Constraints.filter
       (fun (a, _, b) -> Level.Set.mem a uset || Level.Set.mem b uset)
       csts
   in
-  let unames = { quals = [||]; univs = Array.of_list (make_unames univs ounivs)} in
+  let unames = ([||], Array.of_list (make_unames univs ounivs)) in
   let univs = Instance.of_array ([||], Array.of_list univs) in
-  let uctx = UContext.make unames (univs, PConstraints.of_univs csts) in
+  let uctx = UContext.make unames (univs, csts) in
   let subst = snd (make_instance_subst univs) in
   let algs = List.rev_map (subst_univs_level_universe subst) algs in
   (uctx, algs)
@@ -667,11 +658,11 @@ let name_for_core n i =
    should be rare *)
 let name_for n i =
   let base = name_for_core n i in
-  if not (Global.exists_objlabel base) then base
+  if not (Global.exists_objlabel (Label.of_id base)) then base
   else
     (* prevent resetting the number *)
     let base = if i = 0 then base else Id.of_string (Id.to_string base ^ "_") in
-    Namegen.next_global_ident_away (Global.safe_env ()) base Id.Set.empty
+    Namegen.next_global_ident_away base Id.Set.empty
 
 let get_predeclared_ind indn n i =
   if N.equal n (N.append_list N.anon indn) then
@@ -827,7 +818,7 @@ let add_declared n i inst =
 let to_univ_level' u uconv =
   match to_universe uconv.map u with
   | SProp -> (uconv, LSProp)
-  | Type u | GSort (_, u) | VSort (_, u) ->
+  | Type u | QSort (_, u) ->
     let uconv, u = to_univ_level u uconv in
     (uconv, Level u)
   | Set -> (uconv, Level Level.set)
@@ -1007,7 +998,7 @@ let mk_char_list char list char_uinst mkChar s =
 let unfold_proj_case env evd ~field ~indu ~mib ~mip ~args c =
   let ind = fst indu in
   let npar = mib.Declarations.mind_nparams in
-  let ntypes = Declareops.mind_ntypes mib in
+  let ntypes = mib.Declarations.mind_ntypes in
   let u = snd indu in
   let ind_subst =
     List.init ntypes (fun i -> Constr.mkIndU ((fst ind, ntypes - i - 1), u))
@@ -1142,7 +1133,7 @@ let rec to_constr =
               in
               let mib, mip = Inductive.lookup_mind_specif (Global.env()) ind in
               begin
-                match mip.mind_record with
+                match mib.mind_record with
                 | PrimRecord infos ->
                   let p, r =
                     Declareops.inductive_make_projection ind mib ~proj_arg:field
@@ -1401,9 +1392,9 @@ and declare_ind { name = n; params; ty; ctors; univs } i =
         match i with
         | 0 ->
           UContext.make
-            { quals = [||]; univs =  [| Name (Id.of_string "u") |]}
+            ([||], [| Name (Id.of_string "u") |])
             ( Instance.of_array ([||], [| univ_of_name (N.append N.anon "u") |]),
-              PConstraints.empty )
+              Constraints.empty )
         | 1 -> UContext.empty
         | _ -> assert false
       in
@@ -1467,7 +1458,7 @@ and declare_ind { name = n; params; ty; ctors; univs } i =
                        match na.Context.binder_name with
                        | Names.Anonymous -> (ids, (na, t))
                        | Names.Name id ->
-                         let id = Namegen.next_global_ident_away (Global.safe_env ()) id ids in
+                         let id = Namegen.next_global_ident_away id ids in
                          let ids = Id.Set.add id ids in
                          line_msg
                            (N.of_list
@@ -1567,16 +1558,28 @@ and declare_ind { name = n; params; ty; ctors; univs } i =
             List.map
               (fun _ ->
                 {
-                  Record.Data.pf_coercion = None;
-                  pf_instance = None;
+                  Record.Internal.pf_coercion = false;
+                  pf_reversible = false;
+                  pf_instance = false;
+                  pf_priority = None;
+                  pf_locality = Goptions.OptDefault;
                   pf_canonical = false;
                 })
               fields
           in
           let implfs = List.map (fun _ -> []) fields in
+          let projection_fields =
+            List.map
+              (fun (na, t) ->
+                RelDecl.LocalAssum
+                  (EConstr.Unsafe.to_binder_annot na,
+                   EConstr.Unsafe.to_constr t))
+              fields
+          in
           ignore
-            (Record.Internal.declare_projections (mind, 0) ~kind ~inhabitant_id
-               proj_flags implfs)
+            (Record.Internal.declare_projections (mind, 0)
+               (Entries.Polymorphic_entry univs, UnivNames.empty_binders)
+               ~kind inhabitant_id proj_flags implfs projection_fields)
         | _ -> ()
       in
       (mind, algs, ind_name, cnames, univs, squashy)
@@ -1616,14 +1619,14 @@ and declare_ind { name = n; params; ty; ctors; univs } i =
     let inst, uentry =
       let inst = UContext.instance univs in
       let csts = UContext.constraints univs in
-      let { quals = qnames; univs = unames} = UContext.names univs in
+      let qnames, unames = UContext.names univs in
       let uentry =
         match u with
         | LSProp -> UState.Polymorphic_entry univs
         | Level u ->
           UState.Polymorphic_entry
             (UContext.make
-               {quals = qnames; univs = Array.append [| Name (Id.of_string "motive") |] unames}
+               (qnames, Array.append [| Name (Id.of_string "motive") |] unames)
                ( Instance.of_array
                    ([||], Array.append [| u |] (snd (Instance.to_array inst))),
                  csts ))
@@ -1949,8 +1952,8 @@ let do_line state l =
   | None -> do_line ()
   | Some t ->
     (match Control.timeout (float_of_int t) do_line () with
-    | Ok v -> v
-    | Error info -> Exninfo.iraise (TimedOut, info))
+    | Some v -> v
+    | None -> raise TimedOut)
 
 let do_line state l =
   let t0 = System.get_time () in
@@ -1978,8 +1981,8 @@ let do_prefix_line state l =
   | None -> do_prefix_line ()
   | Some t ->
     (match Control.timeout (float_of_int t) do_prefix_line () with
-    | Ok v -> v
-    | Error info -> Exninfo.iraise (TimedOut, info))
+    | Some v -> v
+    | None -> raise TimedOut)
 
 let do_prefix_line state l =
   let t0 = System.get_time () in
